consensus/src/quorum_store/batch_reader.rs

```diff
@@ -346,140 +361,91 @@ impl BatchReader {
+                    .maybe_payload
+                    .expect("Persisted value in QuorumStore DB must have payload");
+                return Ok(payload);
+            },
+            Ok(None) => {
+                unreachable!("Could not read persisted value (according to BatchReader) from DB")
+            },
+            Err(_) => {
+                // TODO: handle error, e.g. from self or not, log, panic.
+            },
+        }
+        Err(Error::CouldNotGetData)
+    }
+
+    pub fn get_batch_from_local(
+        &self,
+        digest: &HashValue,
+    ) -> Result<Vec<SignedTransaction>, Error> {
+        if let Some(value) = self.db_cache.get(digest) {
             if payload_storage_mode(&value) == StorageMode::PersistedOnly {
                 assert!(
                     value.maybe_payload.is_none(),
                     "BatchReader payload and storage kind mismatch"
                 );
-                self.batch_store_tx
-                    .send(BatchStoreCommand::BatchRequest(
-                        *proof.digest(),
-                        self.my_peer_id,
-                        Some(tx),
-                    ))
-                    .await
-                    .expect("Failed to send to BatchStore");
+                self.get_batch_from_db(digest)
             } else {
                 // Available in memory.
-                if tx
-                    .send(Ok(value
-                        .maybe_payload
-                        .clone()
-                        .expect("BatchReader payload and storage kind mismatch")))
-                    .is_err()
-                {
-                    debug!(
-                        "Receiver of requested batch is not available for digest {}",
-                        proof.digest()
-                    );
-                }
+                Ok(value
+                    .maybe_payload
+                    .clone()
+                    .expect("BatchReader payload and storage kind mismatch"))
             }
+        } else {
+            Err(Error::CouldNotGetData)
+        }
+    }
+
+    pub async fn get_batch(
+        &self,
+        proof: ProofOfStore,
+    ) -> oneshot::Receiver<Result<Vec<SignedTransaction>, Error>> {
+        let (tx, rx) = oneshot::channel();
+
+        if let Ok(value) = self.get_batch_from_local(proof.digest()) {
+            tx.send(Ok(value)).unwrap();
         } else {
             // Quorum store metrics
             counters::MISSED_BATCHES_COUNT.inc();
-
-            self.self_tx
-                .send(BatchReaderCommand::GetBatchForSelf(proof, tx))
+            // TODO: handle timeout
+            self.batch_requester
+                .lock()
                 .await
-                .expect("Batch Reader Receiver is not available");
+                .add_request(
+                    *proof.digest(),
+                    proof.shuffled_signers(&self.validator_verifier),
+                    tx,
+                )
+                .await;
```

Comment:
Same comment about the await. Now the execution thread is calling it. Before the execution thread can do other stuff in parallel. Do you think it is significant? For example get data fro the other batches in the meantime.